Add code coverage for bro source files after btest test suite

chungmin99 opened the pull request:
Adds --enable-coverage flag to configure Bro with gcov. A new directory named /testing/bro-code-coverage/ contains a new coverage target that as part of `make coverage` in /testing/. This coverage option creates coverage.log of all important directories in /src/ and places all generated .gcov files alongside the corresponding source file.

ckreibich commented:
Hi Team — could Chung Min and I get some feedback here please? Would be great to be able to finish this. Thanks!

0xxon commented:
Hi,

I actually think this looks pretty great - and I wanted to merge it earlier, but other things came up. In any case - on my first glance I didn't see any big blockers and I hope to get this merged sometime this week.

0xxon commented:
Sorry for the delay - and thanks again for this, this is pretty great.

After taking a look at this, and playing around with it, I encountered a few small issues and was curious if you could fix them. Most of them have to do with portability, which often is a bit of a pain.

* on os-x the code_coverage.sh script currently outputs an error because realpath does not exist:

```bash
$ make
./code_coverage.sh: line 20: realpath: command not found
Checking for coverage files... no .gcda files, nothing to do
```

* on FreeBSD, bash is in a different location. The portable way to call bash is #!/usr/bin/env bash

* could you perhaps rename the directory to testing/coverage? I know that is a question of style, but testing/bro-core-coverage just feels a bit long

* would you mind adding a README to testing/coverage how to read it? I know, that is already somewhat contained in the script file, but having a readme will probably make this more accessible to people looking over it.

* could you add the .gcov file to gitignore? Currently git status looks wild when using code coverage tests.

* make in bro-core-coverage currently fails a bit ungracefully if gcov is not installed on the system. E.g. on FreeBSD, bro can be compiled with coverage but gcov is not installed by default. Would it be possible to make the error message a bit nicer in that case? Current output is:

```bash
$ make
Checking for coverage files... ok
Creating coverage files... ls: /usr/home/johanna/bro/master/testing/bro-core-coverage/tmp.59909/*.gcov: No such file or directory
no gcov files produced, aborting
*** Error code 1

Stop.
make: stopped in /usr/home/johanna/bro/master/testing/bro-core-coverage
```

You do not have to do all of these, especially the system-dependent ones (like if you don't have access to os-x). I would be very thankful if you could address at least a few of these though :)

Thanks again - this is a very nice feature.

0xxon commented:
Also - if you feel like it - I think it would be super neat too if there was a makefile target that outputs the html files that gcov can create.

chungmin99 commented:
Christian, I fixed the issue you just addressed. Function `verify-run` has been replaced successfully with `verify_run`. 

Johanna, I incorporated all the changes you requested/recommended, and checked that it also runs on os-x. Unfortunately, I did not have access to FreeBSD for testing.

0xxon commented:
Thanks a lot - this looks pretty much perfect.
